File: algorithms/cpp/ReverseLinkedList/ReverseLinkedList.cpp
// source: https://leetcode.com/problems/reverse-linked-list/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
	ListNode* reverseList(ListNode* head) {
		if (head == NULL || head->next == NULL) return head;
		ListNode* next = head->next;
		ListNode* new_head = reverseList(head->next);
		next->next = head;
		head->next = NULL;
		return new_head;
	}

	ListNode* reverseList2(ListNode* head) {
		if (head == NULL || head->next == NULL) return head;
		ListNode* node = head->next;
		ListNode* new_head = head;
		while(node != NULL)
		{
			ListNode* temp = node->next;
			node->next = new_head;
			new_head = node;
			node = temp;
		}
		head->next = NULL;
		return new_head;
	}
};

File: courses/linked_list/ReverseLinkedList.java
// https://leetcode.com/problems/reverse-linked-list/

// Reverse a singly linked list. 

// Iterative – O(n) time
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while(curr != null){
            ListNode nextTemp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextTemp;
        }
        return prev;
    }
}

// Recursive – O(n) time
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode p = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return p;
    }
}

